File: examples/adt.rb
def Just(x)
  ->(handler) { handler[:Just].call(x) }
end

Nothing = ->(handler) { handler[:Nothing] }

def match(value, handler)
  value.call(handler)
end

def safeDivide(numerator, denominator)
  denominator == 0 ? Nothing : Just(numerator / denominator)
end

def displayResult(maybeValue)
  extractedValue = match(maybeValue, {
    Just: ->(result) { "Result of division is #{result}" },
    Nothing: "Handled divide by zero error"
  })
  puts extractedValue
end

badDivision = safeDivide(5,0);
goodDivision = safeDivide(22, 7);

displayResult(badDivision);
displayResult(goodDivision);

File: examples/adt.js
const Just = x => handler => handler.Just(x);
const Nothing = handler => handler.Nothing;

const match = (value, handler) => value(handler);

function safeDivide(numerator, denominator) {
    return denominator === 0 ? Nothing : Just(numerator / denominator);
}

function displayResult(mabyeValue) {
  const extractedValue = match(mabyeValue, {
    Just: result => `Result of division is ${result}`,
    Nothing: `Handled divide by zero error`
  });
  console.log(extractedValue);
}

const badDivision = safeDivide(5,0);
const goodDivision = safeDivide(22, 7);

displayResult(badDivision);
displayResult(goodDivision);
